File: 卡码/代码随想录/1/07-回溯/C++/04-排列/LeetCode_0046.cpp
//46. 全排列
#include <iostream>
#include <vector>
using namespace std;

class Solution {
	vector<int> path;
	vector<vector<int>> res;
	vector<bool> visited;
public:
	void backtracking(vector<int>& nums){
		if(path.size()==nums.size()){
			res.push_back(path);
			return;
		}
		for(int i=0;i<nums.size();i++){
			if(visited[i]) continue;
			visited[i]=true;
			path.push_back(nums[i]);
			backtracking(nums);
			path.pop_back();
			visited[i]=false;
		}
	}
	vector<vector<int>> permute(vector<int>& nums) {
		visited.resize(nums.size(),false);
		backtracking(nums);
		return res;
	}
};

int main(){
  system("pause");
  return 0;
}

File: 卡码/代码随想录/1/07-回溯/Js/04-排列/LeetCode_0046.js
//46. 全排列
const permute = function(nums) {
  const res=new Array();
  const path=new Array();
  const visited=new Array(nums.length).fill(false);
  const backtracking=function(nums){
    if(path.length==nums.length){
      res.push(Array.from(path));
      return;
    }
    for(let i=0;i<nums.length;i++){
      if(visited[i]) continue;
      path.push(nums[i]);
      visited[i]=true;
      backtracking(nums);
      visited[i]=false;
      path.pop();
    }
  };
  backtracking(nums);
  return res;
};